### PR TITLE
feat: gracefully exit

### PR DESCRIPTION
**Pull request by skbeh**
Use `exec` on unix and gracefully exit on windows.

### CODE REVIEW

**Review by akiirui — CHANGES_REQUESTED** [review]

File: src/main.rs [hunk 1]
```diff
@@ -3,41 +3,68 @@ mod error;
...
 
-fn main() {
+fn main() -> ExitCode {
     match run() {
-        Ok(_) => (),
-        Err(e) => print_error(e),
+        Some(PlayerExited(err)) => match err {
+            None => ExitCode::FAILURE,
+            Some(code) => ExitCode::from(code),
+        },
```
Comment (akiirui) [inline src/main.rs:19]: Here still exit process directly, try this example `cargo run -- mpv://play/aHR0cHM6Ly92aWRlby5leGFtcGxlLmNvbS8x`

**Comment by akiirui** [comment]
Emm, how do you think about current changes? are they gracefully?

If you want to display mpv exit codes, they look better (based on current dev branch):

```diff
diff --git a/src/error.rs b/src/error.rs
index 0e03c23..dd46c67 100644
--- a/src/error.rs
+++ b/src/error.rs
@@ -12,8 +12,10 @@ pub enum Error {
     DangerousVideoProtocol(String),
     #[error("Failed to get config directory")]
     FailedGetConfigDir,
-    #[error("Player exited with error or termination signal")]
-    PlayerExited,
+    #[error("Player exited by error")]
+    PlayerExited(u8),
+    #[error("Player exited by signal")]
+    PlayerExitedBySignal,
     #[error("Failed to run player. {0}")]
     PlayerRunFailed(std::io::Error),
     #[error("Failed to decode ({0})")]
diff --git a/src/main.rs b/src/main.rs
index 2dc8cba..21ced33 100644
--- a/src/main.rs
+++ b/src/main.rs
@@ -3,14 +3,16 @@ mod error;
 mod plugins;
 mod protocol;
 
+use std::process::ExitCode;
+
 use crate::config::Config;
 use crate::error::Error;
 use crate::plugins::Plugins;
 use crate::protocol::Protocol;
 
-fn main() {
+fn main() -> ExitCode {
     match run() {
-        Ok(_) => (),
+        Ok(_) => ExitCode::SUCCESS,
         Err(e) => print_error(e),
     }
 }
@@ -26,6 +28,7 @@ fn run() -> Result<(), Error> {
 
     match arg {
         "-v" | "--version" => Ok(print_usage()),
+
         _ => {
             let proto = Protocol::parse(arg)?;
             let config = Config::load()?;
@@ -47,8 +50,12 @@ fn print_usage() {
 }
 
 /// Print error
-fn print_error(e: Error) {
-    eprint!("ERROR: {}", e);
-    std::io::Read::read(&mut std::io::stdin(), &mut [0]).unwrap();
-    std::process::exit(1);
+fn print_error(e: Error) -> ExitCode {
+    eprint!("{e}");
+    std::io::Read::read(&mut std::io::stdin(), &mut []).unwrap();
+
+    match e {
+        Error::PlayerExited(code) => ExitCode::from(code),
+        _ => ExitCode::FAILURE,
+    }
 }
diff --git a/src/plugins/play.rs b/src/plugins/play.rs
index f929372..d4c59d7 100644
--- a/src/plugins/play.rs
+++ b/src/plugins/play.rs
@@ -45,7 +45,7 @@ pub fn exec(proto: &Protocol, config: &Config) -> Result<(), Error> {
             cookies_option.push_str(&p.display().to_string());
             options.push(&cookies_option);
         } else {
-            eprintln!("Cookies file {} doesn't exist", v);
+            eprintln!("Cookies file {v} doesn't exist");
         }
     }
 
@@ -87,9 +87,12 @@ pub fn exec(proto: &Protocol, config: &Config) -> Result<(), Error> {
         .status();
 
     match player {
-        Ok(o) => match o.success() {
-            true => Ok(()),
-            false => Err(Error::PlayerExited),
+        Ok(o) => match o.code() {
+            Some(code) => match code {
+                0 => Ok(()),
+                _ => Err(Error::PlayerExited(code as u8)),
+            },
+            None => Err(Error::PlayerExitedBySignal),
         },
         Err(e) => Err(Error::PlayerRunFailed(e)),
     }
```

**Comment by skbeh** [comment]
I don't want to display mpv exit codes. I just want to avoid the usage of `std::process::exit` and let the exit code match mpv's.

**Comment by akiirui** [comment]
> I don't want to display mpv exit codes. I just want to avoid the usage of `std::process::exit` and let the exit code match mpv's.

Yep these codes in previous reply will let the exit code match mpv's.

File: src/main.rs [hunk 2]
```diff
@@ -3,29 +3,54 @@ mod error;
...
 
-fn main() {
-    match run() {
-        Ok(_) => (),
-        Err(e) => print_error(e),
-    }
+fn main() -> ExitCode {
+    let exit_code = match run() {
+        Err(PlayerExited(err)) => match err {
+            None => ExitCode::from(4),
```
Comment (akiirui) [inline src/main.rs:17]: If process is terminated by different signal, the exit code doesn't always equal 4 such as `SIGKILL (9)`.

**Comment by akiirui** [comment]
The signal is sent by users, They know what they've done, so I assume the exit code from the signal is success.

So, only match the exitcode from mpv is fine.

Thanks for your work.

**Pull request merged**